### ISSUE TITLE
Make HttpUriFetcher / HttpUrlFetcher public

### ISSUE THREAD

**Issue by Tolriq** [coil-kt/coil#883]
**Is your feature request related to a problem? Please describe.**
HttpUriFetcher / HttpUrlFetcher are internal and can't be reused in custom fetchers

**Describe the solution you'd like**
As per the title make HttpUriFetcher / HttpUrlFetcher public so I can use them in a custom fetcher with dynamic okhttp clients.

**Additional context**
My application can be connected at the same time to multiple different "hosts" that all provides their custom okhttp clients with the necessary authentification / oauth / ... interceptors.
I need to be able to download the images from them via coil. 

I can easily create a custom fetcher to get the proper ```Call.Factory``` and urls but can't reuse what is currently provided by coil and would require to copy paste hoping that there won't be any other internal things used.

Edit: Actually await and getMimeTypeFromUrl are internal :(

Edit2: FileFetcher is very simple but it could be useful too when handling the cache ourselves

**Comment by colinrtwhite** [comment]
Duplicate of https://github.com/coil-kt/coil/issues/637. Delegating to internal fetchers is already implemented in 2.0, which is intended to be out soon (though no ETA).

**Comment by Tolriq** [comment]
@colinrtwhite 

This is not the same need at all.

I do not want to, nor can, delegate to default fetchers as I need to pass custom okhttp to each fetch that are driven by the data. 

I have a custom fetcher for a custom data type. The fetcher is injected with a component that return the proper okhttp instance needed for that particular data and then should instantiate the HttpFetcher with the proper client.

Basically with a copy paste of the code and all necessary internals the fecth part can be resumed to :
```
return HttpUriFetcher(data.getOkHttpClient()).fetch(pool, Uri.parse(data.getUrl()), size, options)
```

But I need to duplicate a lot of your code for no real benefits.

If `HttpFetcher` was public I could easily provide an `HttpDataFetcher() : HttpFetcher ` to fit that need without any hassle.

**Comment by colinrtwhite** [comment]
@Tolriq Gotcha - I'm a bit apprehensive to make `HttpFetcher` public because it restricts the changes we can make to those classes in the future once it's public API. I don't think it's likely it's made public. To support dynamic OkHttp clients your best bet is to provide a custom `Call.Factory` that swaps between the `OkHttp` clients internally.

**Comment by Tolriq** [comment]
@colinrtwhite I can't do that as the data is not available when it reaches it. Coil already only have an HttpUrl when it calls it.

While I can copy the file to reuse it, it spreads to some other internal call making me copy more than just one file. And I fear that at any time like 2.0 it start using more internal things spread everywhere and that I can no more do that and I'm forced to reinvent the wheel :(

**Comment by David-Development** [comment]
@Tolriq May I ask what you ended up doing? I'm facing the same issue right now. I need to parse some additional data from the response and wanted to simply extend the existing HttpFetcher
